packet/routes/api.py

```diff
@@ -1,19 +1,31 @@
-from packet import auth, app
-from packet.utils import before_request
-from packet.packet import sign as sign_packet
+from packet import app, db
+from packet.utils import before_request, packet_auth
+from packet.models import Packet, MiscSignature
 
 
-@app.route("/api/v1/<member_username>/sign/<packet_username>", methods=["POST"])
-@auth.oidc_auth
+@app.route("/api/v1/sign/<packet_id>/", methods=["POST"])
+@packet_auth
 @before_request
-def sign(member_username, packet_username, info):
-    if info['uid'] != member_username:
-        if info['member_info']:
-            if "eboard-evaluations" not in info['member_info']['group_list']:
-                return "Error: You are not evals"
+def sign(packet_id, info):
+    packet = Packet.by_id(packet_id)
+
+    if packet is not None and packet.is_open():
+        if app.config["REALM"] == "csh":
+            # Check if the CSHer is an upperclassman and if so, sign that row
+            for sig in filter(lambda sig: sig.member == info["uid"], packet.upper_signatures):
+                sig.signed = True
+                db.session.commit()
+                return "Success: Signed Packet: " + packet.freshman_username
+
+            # The CSHer is a misc so add a new row
+            db.session.add(MiscSignature(packet=packet, member=info["uid"]))
+            db.session.commit()
+            return "Success: Signed Packet: " + packet.freshman_username
         else:
-            return "Error: UID Submission Mismatch"
-    if not sign_packet(member_username, packet_username):
-        return "Error: Signature not valid.  Reason: Unknown"
-    return "Success: Signed Packet: " + packet_username
+            # Check if the freshman is onfloor and if so, sign that row
+            for sig in filter(lambda sig: sig.freshman_username == info["uid"], packet.fresh_signatures):
+                sig.signed = True
+                db.session.commit()
+                return "Success: Signed Packet: " + packet.freshman_username
 
+    return "Error: Signature not valid.  Reason: Unknown"
```

Comment:
This should return a more detailed error message in a parsable JSON response with an explicit status code, e.g.

```
HTTP/1.1 403 Forbidden
Content-Type: application/json

{
    "status": "error",
    "errors": [
        {
            "title":  "Invalid Signature",
            "detail": "You do not have permission to sign this packet."
        }
    ]
}
```

or

```
HTTP/1.1 400 Bad Request
Content-Type: application/json

{
    "status": "error",
    "errors": [
        {
            "title":  "Invalid Signature",
            "detail": "You have already signed this packet."
        }
    ]
}
```

Reply:
I'll be sure to include this when we switch over to REST.